UMLEditor/UMLEditor/Classes/TimeMachine.cs

```diff
@@ -0,0 +1,134 @@
+﻿namespace UMLEditor.Classes;
+
+/// <summary>
+/// Class to represent the action history of the UML program
+/// </summary>
+public static class TimeMachine
+{
+    /// <summary>
+    /// Node class to represent a single state in the time machine history.
+    /// </summary>
+    public class Node
+    {
+        /// <summary>
+        /// Reference to previous node
+        /// </summary>
+        public Node? PrevNode;
+        /// <summary>
+        /// Diagram at this node
+        /// </summary>
+        public Diagram? StateDiagram;
+        /// <summary>
+        /// Reference to next node
+        /// </summary>
+        public Node? NextNode;
+    }
+    /// <summary>
+    /// References to the beginning, end, and current state of the time machine
+    /// </summary>
+    public static Node? Head, Tail, Current;
+    /// <summary>
+    /// Counter for number of states in the time machine
+    /// </summary>
+    public static int? Size;
+
+    /// <summary>
+    /// Initializes the time machine to an empty list
+    /// </summary>
+    public static void Initialize()
+    {
+        Head = null;
+        Tail = Head;
+        Current = Tail;
+        Size = 0;
+    }
+
+    /// <summary>
+    /// Adds a new diagram to the end of the list. Erases redo history if diagram is added to middle of list.
+    /// </summary>
+    /// <param name="d">Diagram to add</param>
+    public static void AddState(Diagram d)
+    {
+        // If in the middle of the list clears out all future nodes
+        if (Current?.NextNode is not null)
+        {
+            Node? temp = Current.NextNode;
+            while (temp is not null)
+            {
+                Node? next = temp.NextNode;
+                temp.StateDiagram = null;
+                temp.PrevNode = null;
+                temp.NextNode = null;
+                temp = next;
+                Size--;
+            }
+        }
+        
+        // Add new node to end of list
+        Node newNode = new Node();
+        newNode.StateDiagram = new Diagram(d);
+        newNode.PrevNode = Current;
+        newNode.NextNode = null;
+        if (Current is not null)
+        {
+            Current.NextNode = newNode;
+        }
+        Tail = newNode;
+        Size++;
+        Current = Tail;
+        if (Size == 1)
+        {
+            Head = newNode;
+        }
+    }
+
+    /// <summary>
+    /// Moves to next state in the time machine (redo)
+    /// </summary>
+    /// <returns>The next diagram, or null if there are none</returns>
+    public static Diagram? NextState()
```

Comment:
The return value from NextState (and PreviousState) should not be null. While a null return value can be used to indicate that such a state does not exist, it does not make sense to be doing that here. With how this is now, I cannot ask the TimeMachine the question _is there a next state?_ without potentially changing the state of the time machine (performing an undo and redo). According to the user stories, we want to have the undo and redo buttons be greyed out. We need a way to be able to test if there is a state to go back or forward on without needing to change the state of the TM. We want the buttons to be greyed out when there are no such states- a user should not have to click the button first to realize we ran out of states.